File: 购物车命令模式/购物车命令模式/DynamicCommand.m
//
//  DynamicCommand.m
//  购物车命令模式
//
//  Created by qrh on 2019/7/23.
//  Copyright © 2019 zdq. All rights reserved.
//

#import "DynamicCommand.h"

@interface DynamicCommand ()

@property (nonatomic, strong) ShoppingReceive * receiver;
@property (nonatomic, copy) DynamicBlock  block;

@end

@implementation DynamicCommand

- (instancetype)initReceiver:(ShoppingReceive *)receiver block:(DynamicBlock)block;
{
    if (self = [super init]) {
        self.receiver = receiver;
        self.block = block;
    }
    return self;
}

// 类方法
+(id<CommonProtocol>)createCommand:(ShoppingReceive *)receiver block:(DynamicBlock)block;
{
    return [[DynamicCommand alloc]initReceiver:receiver block:block];
}

- (void)excute;
{
    self.block(self.receiver);
}

@end

File: 命令模式/命令模式/Dynamic/DynamicCommand.swift
//
//  DynamicCommand.swift
//  命令模式
//
//  Created by qrh on 2019/7/19.
//  Copyright © 2019 zdq. All rights reserved.
//

import Foundation


class DynamicCommand: TMCommandProtocol {
    private var tm : TetrisMachine
    private var block: (TetrisMachine)-> Void
    
    init(tm: TetrisMachine, block: @escaping(TetrisMachine)-> Void) {
        self.tm = tm
        self.block = block
    }
    
    func excute() {
        self.block(self.tm)
    }
    
    class func creatCommand(tm: TetrisMachine, block: @escaping(TetrisMachine)-> Void) -> TMCommandProtocol {
        return DynamicCommand(tm: tm, block: block)
    }
}
